GVFS/GVFS.Common/Prefetch/BlobPrefetcher.cs

```diff
@@ -129,6 +149,44 @@ public static bool TryLoadFileList(Enlistment enlistment, string filesInput, Lis
             return true;
         }
 
+        public static bool IsNoopPrefetch(
+            ITracer tracer,
+            FileBasedDictionary<string, string> lastPrefetchArgs,
+            string commitId,
+            List<string> files,
+            List<string> folders,
+            bool hydrateFilesAfterDownload)
+        {
+            if (lastPrefetchArgs != null &&
+                lastPrefetchArgs.TryGetValue(PrefetchArgs.CommitId, out string lastCommitId) &&
+                lastPrefetchArgs.TryGetValue(PrefetchArgs.Files, out string filesString) &&
+                lastPrefetchArgs.TryGetValue(PrefetchArgs.Folders, out string foldersString) &&
+                lastPrefetchArgs.TryGetValue(PrefetchArgs.Hydrate, out string hydrateString))
+            {
+                bool isNoop =
+                    lastCommitId == commitId &&
+                    hydrateString == hydrateFilesAfterDownload.ToString() &&
+                    JsonConvert.SerializeObject(files) == filesString &&
+                    JsonConvert.SerializeObject(folders) == foldersString;
+
+                tracer.RelatedEvent(
+                    EventLevel.Informational,
+                    "BlobPrefetcher.IsNoopPrefetch", 
+                    new EventMetadata
+                    {
+                        { PrefetchArgs.CommitId, lastCommitId },
+                        { PrefetchArgs.Files, filesString },
+                        { PrefetchArgs.Folders, foldersString },
+                        { PrefetchArgs.Hydrate, hydrateString },

```

Comment:
Also trace the new values?

Reply:
Maybe log all of them when `"Result"` is false?  My main concern was being able to track down why prefetch might have decided that the operation was not a no-op.

In addition to the "expanded paths out of a file" scenario, we're also not logging the new commit ID anywhere.

Reply:
I suspect more about false positives rather than false negatives. I just went ahead and logged it in all cases.

e.g. for the command
```
Saeeds-MacBook-Pro:GVFSTest sanoursa$ ~/Repos/VFSForGit/Publish/gvfs prefetch --folders 'GVFS;ProjFS.Mac;MirrorProvider' --hydrate
All requested files are already available. Nothing new to prefetch.
```

Got the log:
```
[2018-11-28 14:59:28 -08:00] Prefetch (Start) {"Version":"0.2.173.2","EnlistmentRoot":"/Users/sanoursa/GVFSTest","Remote":"https://dev.azure.com/gvfs/ci/_git/ForTests","ObjectsEndpoint":"https://dev.azure.com/gvfs/ci/_git/ForTests"}
[2018-11-28 14:59:28 -08:00] PerformPrefetch {"Commits":false,"Files":"","Folders":"GVFS;ProjFS.Mac;MirrorProvider","FoldersListFile":"","HydrateFiles":true}
[2018-11-28 14:59:29 -08:00] BlobPrefetcher.IsNoopPrefetch {"LastCommitId":"2797fbb8358bb2e0c12d6f3b42a60b43f7655edf","LastFiles":"[]","LastFolders":"[\"/Users/sanoursa/GVFSTest/src/GVFS/\",\"/Users/sanoursa/GVFSTest/src/ProjFS.Mac/\",\"/Users/sanoursa/GVFSTest/src/MirrorProvider/\"]","LastHydrate":"True","NewCommitId":"2797fbb8358bb2e0c12d6f3b42a60b43f7655edf","NewFiles":"[]","NewFolders":"[\"/Users/sanoursa/GVFSTest/src/GVFS/\",\"/Users/sanoursa/GVFSTest/src/ProjFS.Mac/\",\"/Users/sanoursa/GVFSTest/src/MirrorProvider/\"]","NewHydrate":"True","Result":true}
[2018-11-28 14:59:29 -08:00] Prefetch (Stop) {"DurationMs":615}
```

Reply:
> I suspect we'll care more about false positives rather than false negatives. I just went ahead and logged it in all cases.

That's a good point, I agree false positives would be worse than false negatives.